TC-G-2.3: Automate

gvargas-csa opened the pull request:
Add new automated test case TC_G_2.3 https://github.com/project-chip/matter-test-scripts/issues/383

#### Testing

Tested on the following apps:

darwin-x64-all-clusters-no-ble/chip-all-clusters-app

gvargas-csa committed:
chore(TC_G-2_3): replace all steps format use TC-G-2_2 suggestions

gvargas-csa committed:
Merge branch 'master' into HEAD

gvargas-csa committed:
chore(TC-G-2.3): all steps done

gvargas-csa committed:
chore(TC-G-2.3): restyled by autopep8

gvargas-csa committed:
chore(TC-G-2.3): restyled by isort

gvargas-csa committed:
chore(TC-G-2.3): restyled by autopep8

gvargas-csa committed:
chore(TC-G-2.3): restyled by autopep8

gvargas-csa committed:
Merge branch 'master' into HEAD

gvargas-csa committed:
Merge branch 'master' into HEAD

gvargas-csa committed:
chore(TC-G-2.3): remove hardcodes values and replaced endpoints

gvargas-csa committed:
chore(TC-G-2.3): restyled by autopep8

gvargas-csa committed:
Merge branch 'master' into HEAD

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+                TestStep("1c", "TH cleans up the groups by sending the RemoveAllGroups command to the DUT on PIXIT.G.ENDPOINT"),
+                TestStep("1d", "TH sends AddGroup command to DUT on PIXIT.G.ENDPOINT as unicast with the following fields : 1. GroupID as 0x0002 2. GroupName as Gp2"),
+                TestStep("2a", "TH reads GroupTable attribute from the GroupKeyManagement cluster from DUT on EP0"),
+                TestStep("2b", "Verify that the GroupTable contains an entry with the GroupName as Gp2"),
```
and below for all the other GN ones.

gvargas-csa replied:
I did for all ones too. Thank you!

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        asserts.assert_equal(resp[0].Status, Status.Success, "GroupKeyMap attribute write failed")
+
+        self.step("9b")
+        result = await th1.SendCommand(self.dut_node_id, self.matter_test_config.endpoint, Clusters.Identify.Commands.Identify(0x0078))
```
That looks like all the steps through the end of the test so you should be able to just early-exit here if there's no identify cluster (I think - worth double checking)

gvargas-csa replied:
I reviewed each step individually  instead of skipping all remaining steps at once. I think this approach is more consistent with the rest of the test case. Do you think that's okay?

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        except Exception as e:
+            logging.exception("Error while sending command AddGroupIfIdentify %s", e)
+
+        groupTableList: List[Clusters.GroupKeyManagement.Attributes.GroupTable] = await self.read_single_attribute(
```
why is this read here?

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+            dev_ctrl=th1, node_id=self.dut_node_id, endpoint=0, attribute=Clusters.GroupKeyManagement.Attributes.GroupTable)
+        asserts.assert_true(all(entry.groupId != kGroupId for entry in groupTableList),
+                            f"Group ID {kGroupId} was unexpectedly found in the group table")
+
```
Does this test leave extra groups on the device? 

If so, should add a cleanup step.

gvargas-csa replied:
By design, if the test runs to the end it does not leave extra groups. Step 18 clears all groups, and step 23 checks that the group added in step 22 is not in the table. If the test fails before that, a group may stay, but I think this is normal since we need to see the error case.

gvargas-csa committed:
Merge branch 'master' into HEAD

gvargas-csa committed:
Update src/python_testing/TC_G_2_3.py

Added conditional wording to all steps related to the Group Name (GN) feature, based on Cecille’s suggestion.

Co-authored-by: C Freeman <cecille@google.com>

gvargas-csa committed:
Addressed PR feedback from Cecille's suggesstions:

- Added `@run_if_endpoint_matches(Clusters.Groups)` decorator to ensure tests run on the correct Groups cluster endpoint, as suggested.
- Replaced references to `PIXIT.G.ENDPOINT` with "Groups cluster endpoint" in test step descriptions, since the PIXIT value is not actually used. We might also consider updating the test plan to reflect this change.
- Added conditional wording to all steps related to the Group Name (GN) feature, based on Cecille’s suggestion.
- Added a second assertion to verify that the returned Group ID is 0x0002.
- Removed unused attribute read that was being repeated later in the code.

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        return [TestStep(0, "Comissioning, already done", is_commissioning=True),
+                TestStep("1a", "TH sends KeySetWrite command in the GroupKeyManagement cluster to DUT on EP0 using a key that is pre-installed on the TH. GroupKeySet fields are as follows:"),
+                TestStep("1b", "TH writes the GroupKeyMap attribute in the GroupKeyManagement cluster of DUT on EP0 with maxgroups entries binding GroupId(0x0002 to (maxgroups+1)) with GroupKeySetID 1"),
+                TestStep("1c", "TH cleans up the groups by sending the RemoveAllGroups command to the DUT on PIXIT.G.ENDPOINT"),
```
Thanks! I replaced the `PIXIT.G.ENDPOINT` references with "Groups cluster endpoint" to better reflect how the test actually behaves. We might want to update the test plan to match, what do you think?

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        self.step("1d")
+        kGroupId2 = 2
+        kGroupNameGp2 = "Gp2"
+        groupTableList: List[Clusters.GroupKeyManagement.Attributes.GroupTable] = await self.read_single_attribute(
```
I think left in by mistake, I've removed it now.

gvargas-csa committed:
Restyled by autopep8

gvargas-csa committed:
Refactor: Replace manual GroupName feature check with feature_guard

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        group_name_supported: bool = (feature_map & (1 << name_feature_bit)) != 0
+
+        self.step("2b")
+        if group_name_supported:
```
Thanks! I updated the logic to use `feature_guard` for the GroupName feature check. Much cleaner now, let me know if it looks good to you!

gvargas-csa committed:
Add assertion and minor formatting cleanup

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        self.step("3")
+        kGroupId3 = 3
+        kGroupNameGp3 = "Gp3"
+        groupTableList: List[Clusters.GroupKeyManagement.Attributes.GroupTable] = await self.read_single_attribute(
```
Yeah, it was a copy/paste thing. I use groupTableList a lot in different steps, and during some refactoring I ended up leaving that extra call by mistake. I’ve cleaned it up now. Thanks for pointing it out!

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        asserts.assert_true(found, f"groupId {kGroupId3} not found in GroupTable")
+
+        self.step("4b")
+        if group_name_supported:
```
Thanks! I updated the logic to use feature_guard for the GroupName feature check

gvargas-csa committed:
Remove unused addGroup command for Group 6

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+
+        kGroupId6 = 6
+        kGroupNameGp6 = "Gp6"
+        result = await th1.SendCommand(self.dut_node_id, self.matter_test_config.endpoint, Clusters.Groups.Commands.AddGroup(kGroupId6, kGroupNameGp6))
```
I initially added the `AddGroupIfIdentifying` command, but I didn't work properly with the endpoint I was using. So I tried using `AddGroup` manually to see if that would work. Later, I went back to using `AddGroupIfIdentifying`, but I forgot to remove the old lines. I've cleaned it up now.

gvargas-csa committed:
remove old lines add new check for cluster and attributes

gvargas-csa committed:
Merge branch 'project-chip:master' into TC-G-2.3

gvargas-csa committed:
Merge branch 'master' into TC-G-2.3

gvargas-csa committed:
Fix: rename python namespace to matter* instead chip*

gvargas-csa committed:
Patch: Restyled by isort

gvargas-csa committed:
TC-Groups 2.3: instead logging use assert for test failure

gvargas-csa commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,397 @@
+        self.step("12")
+        kGroupId7 = 7
+        kGroupNameGp7 = "Gp7"
+        try:
```
The command it works it just is not returned the response.

gvargas-csa committed:
Remove unused import logging

gvargas-csa committed:
Groups 2.3: Fix AddGroupIfIdentifying test to just send command without response

gvargas-csa committed:
Groups 2.3: minor cleanup correct typo in variable name

gvargas-csa committed:
Groups 2.3: updated step 15 to just expect success

gvargas-csa committed:
Groups 2.3: Add suggestions

- In verify_group_membership_response, replace Nullable.
- Remove unnecessary int() cast, since the value is already int.
- Step 13: verify both groupId and endpoint are correct.

gvargas-csa committed:
Merge branch 'master' into TC-G-2.3

gvargas-csa committed:
Groups 2.3: calculate differentGroupId dynamically instead of hardcoding

gvargas-csa committed:
Merge branch 'master' into TC-G-2.3

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,382 @@
+                    break
+            asserts.assert_true(group_found, f"Group with name '{kGroupNameGp3}' not found in the GroupTable")
+        else:
+            self.mark_current_step_skipped()
```
```suggestion
```

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,382 @@
+        self.step("9b")
+        if await self.command_guard(endpoint=self.matter_test_config.endpoint, command=Clusters.Identify.Commands.Identify):
+            result = await th1.SendCommand(self.dut_node_id, self.matter_test_config.endpoint, Clusters.Identify.Commands.Identify(0x0078))
+        else:
```
please remove all of these where you've used a guard.

cecille commented on src/python_testing/TC_G_2_3.py:
```diff
@@ -0,0 +1,382 @@
+
+        self.step("16a")
+        GroupKeyMapStruct = Clusters.GroupKeyManagement.Structs.GroupKeyMapStruct
+        groupKeyMapStruct = [GroupKeyMapStruct(groupId=i, groupKeySetID=kGroupKeySetID) for i in range(6+maxgroups, 6+maxgroups+1)]
```
how is this working out? range is not inclusive of the end (https://docs.python.org/2/library/functions.html#range) so this is setting a single value - 6+maxgroups.

so how are you getting resource exhaused? Also, 6+maxgroups below is not a different group ID

Can you please check the logic in this test and ensure things are working as expected? Is this related to the incorrect range in step 1b?

cecille reviewed (commented):
Please run this test against the following app setups to ensure it works

1) device endpoint with groupname feature enabled, and identify cluster present on endpoint
2) device endpoint with groupname feature NOT enabled, identify cluster present on endpoint
3) device endpoint with groupname feature enabled, and identify cluster NOT present on endpoint
4) device endpoint with groupname feature NOT enabled, and identify cluster NOT present on endpoint
